### DISCUSSION TITLE
Slow work in unity

### DISCUSSION THREAD

**Discussion by gosodin** [SixLabors/ImageSharp#1754] [Q&A]
Hello, there was a need to process bitmaps in unity, as well as work with exif orientation, and your library is just super. But I ran into a rather strange and big problem - very slow work. What a normal console application does in seconds, in unity it is done in minutes. Most likely, these are some limitations of the platform itself, but I am writing in the hope that there is some solution to this problem.

**Comment by brianpopow** [comment]
A lot of performance from ImageSharp comes from using Hardware Intrinsics, but those are not available with unity, which explains the slowness you see. I am afraid, that there is not much we can do about it. 

I have seen that Unity started to support Hardware Intrinsics with [Burst](https://docs.unity3d.com/Packages/com.unity.burst@0.2/manual/index.html), but as far as i understand this, this does not apply to third party library's like ours.

**Reply by antonfirsov** [reply]
> What a normal console application does in seconds, in unity it is done in minutes.

Looking at the scale of the problem, I'm not sure if it's really about SIMD. @gosodin do you use IL2CPP or mono "scripting backend"?

**Reply by gosodin** [reply]
> > What a normal console application does in seconds, in unity it is done in minutes.
> 
> Looking at the scale of the problem, I'm not sure if it's really about SIMD. @gosodin do you use IL2CPP or mono "scripting backend"?

mono

**Comment by antonfirsov** [comment]
@gosodin I remember cases where mono based JIT was falling back to interpreted mode on performance critical ImageShap code paths. You need IL2CPP to get reasonable perf. Let us know if changing the scripting backend helps!

**Reply by gosodin** [reply]
> @gosodin I remember cases where mono based JIT was falling back to interpreted mode on performance critical ImageShap code paths. You need IL2CPP to get reasonable perf. Let us know if changing the scripting backend helps!

so far, I just changed the scripting backend and got an exception:

ExecutionEngineException: Attempting to call method 'SixLabors.ImageSharp.Processing.Processors.Transforms.CropProcessor::CreatePixelSpecificCloningProcessor<SixLabors.ImageSharp.PixelFormats.Rgba32>' for which no ahead of time (AOT) code was generated.
08-27 15:05:35.934 17156 17179 E Unity : at SixLabors.ImageSharp.Processing.Processors.CloningImageProcessor.SixLabors.ImageSharp.Processing.Processors.IImageProcessor.CreatePixelSpecificProcessor[TPixel] (SixLabors.ImageSharp.Configuration configuration, SixLabors.ImageSharp.Image`1[TPixel] source, SixLabors.ImageSharp.Rectangle sourceRectangle) [0x00000] in <00000000000000000000000000000000>:0
08-27 15:05:35.934 17156 17179 E Unity : at SixLabors.ImageSharp.Processing.DefaultImageProcessorContext`1[TPixel].ApplyProcessor (SixLabors.ImageSharp.Processing.Processors.IImageProcessor processor, SixLabors.ImageSharp.Rectangle rectangle) [0x00000] in <00000000000000000000000000000000>:0
08-27 15:05:35.934 17156 17179 E Unity : at SixLabors.ImageSharp.Processing.DefaultImageProcessorContext`1[TPixel].ApplyProcessor (SixLabors.ImageSharp.Processing.Processors.IImageProcessor pr

I will continue to understand the problem a little later

**Reply by antonfirsov** [reply]
@gosodin yeah IL2CPP AOT needs some tweaking unfortunately, this could be helpful: https://github.com/SixLabors/ImageSharp/issues/1703#issuecomment-896900448
